server/services/ai.js

```diff
@@ -2,24 +2,45 @@ const FIRST_TICK_BULK_UPGRADE_SCI_PERCENTAGE = 20;
 const FIRST_TICK_BULK_UPGRADE_IND_PERCENTAGE = 30;
 const LAST_TICK_BULK_UPGRADE_ECO_PERCENTAGE = 100;
 
-module.exports = class AIService {
+const Heap = require('qheap');
+const { getOrInsert, minBy, minElementBy, reverseSort } = require('../utils.js')
+
+const DEFEND_STAR_ACTION = 'DEFEND_STAR';
+const CLAIM_STAR_ACTION = 'CLAIM_STAR';
+const REINFORCE_STAR_ACTION = 'REINFORCE_STAR';
 
-    constructor(starUpgradeService) {
+module.exports = class AIService {
+    constructor(starUpgradeService, carrierService, starService, distanceService, waypointService, combatService) {
         this.starUpgradeService = starUpgradeService;
+        this.carrierService = carrierService;
+        this.starService = starService;
+        this.distanceService = distanceService;
+        this.waypointService = waypointService;
+        this.combatService = combatService;
     }
 
     async play(game, player) {
```

Comment:
Currently the full game gets passed into this function, meaning a game where all stars, all carriers, all everything is basically visible. Which seems weird, as the AI is meant to be just like a player, but then automated.

I'd suggest using the this.gameGalaxyService.getGalaxy(gameId, userId, tick) to get everything in the vision on the AI.
Working with this is not only more logical, as it is all the information the AI should know, it is also faster, as it is a game with only the visible stars and carriers, and therefore a lot less irrelevant junk